# Proteomics data: caveats

```{r env, echo=FALSE}
suppressPackageStartupMessages(library("Pbase"))
```

**Mapping** of *peptides along protein sequences (although not
  explicitly considered a mapping task)* and *short reads along genome
  coordinates*.

But...

## Protein inference

![Basic peptide grouping](./figures/F5.large.jpg)

From
[Nesvizhskii and Aebersold (2005)](http://www.ncbi.nlm.nih.gov/pubmed/16009968).


![Peptide evidence classes](./figures/nbt0710-647-F2.gif)

From [Qeli and Ahrens (2010)](http://www.ncbi.nlm.nih.gov/pubmed/20622826)

Often, in proteomics experiments, the features represent single
proteins and **groups of indistinguishable proteins** identified by
shared (non-unique) peptides.

## Mapping protein and gene identifers

The protein database and the genome are _independent_, i.e. the
proteins do not make explicitly reference to the genome they originate
from.

![DB in proteomics](./figures/indep-prot-db.png)

![linking with genomics](./figures/gen-prot-db.png)

If we want to map UniProt accession to genomic identifiers (Ensembl
transcipt identifiers)

```{r ids, cache=TRUE}
## The UniProt human proteome (release 2015_02)
library("Pbase")
up <- Proteins("data/HUMAN_2015_02.fasta.gz")
length(up)

## Using the accession number to Ensembl Biomart query
## for transcript identifiers
library("biomaRt")
ens <- useMart("ensembl", "hsapiens_gene_ensembl")
ens
upbm <- select(ens, keys = seqnames(up),
               keytype = "uniprot_swissprot_accession",
               columns = c(
                   "uniprot_swissprot_accession",
                   "ensembl_transcript_id"))
```

```{r ids1}
## How many UniProt accession with Ensembl transcripts were found?
table(seqnames(up) %in% unique(upbm$uniprot_swissprot_accession))
```

```{r ids2}
## How many transcripts per accession do we find?
table(table(upbm$uniprot_swissprot_accession))
```

## Coverage

**coverage** in proteomics in `%` and RNA-Seq `X`

```{r cvg}
cvg <- data.table::fread("./data/Ensembl_76.csv", skip = 17,
                         stringsAsFactors = FALSE)
summary(cvg$coverage)
```

This has an impact on **protein inference** and **missing values**.

## Missing values

 There are two types of mechanisms resulting in missing values in
 LC/MSMS experiments.

- Missing values resulting from absence of detection of a feature,
  despite ions being present at detectable concentrations.  For
  example in the case of ion suppression or as a result from the
  stochastic, data-dependent nature of the MS acquisition
  method. These missing value are expected to be randomly distributed
  in the data and are defined as **missing at random** (MAR) or
  **missing completely at random** (MCAR).

- Missing values resulting from the absence of the low abundance of
  ions (below the limit of detection of the instrument). These missing
  values are not expected to be randomly distributed in the data and
  are defined as **missing not at random** (MNAR).

MNAR features should ideally be imputed with a **left-censor**
method. Conversely, it is recommended to use **hot desk** methods when
data are missing at random.
