# Mapping peptides to genomic coordinates

Illustration with [`Pbase`](http://bioconductor.org/packages/devel/bioc/html/Pbase.html).

```{r env, echo=FALSE}
suppressPackageStartupMessages(library("Pbase"))
library("knitr")
data(p)
```

We have an example data composed of `r length(p)` proteins, with
UniProt accession numbers and Ensembl transcipt identifers and each
protein has a set experimentally observed peptides:

```{r ids, echo=FALSE}
kable(cbind(Acc = seqnames(p),
            ENST = acols(p)$ENST,
            npep = elementLengths(pranges(p))),
      row.names = 0)
```

For example, `r seqnames(p)[5]`:

```{r pplot1, echo=FALSE}
plot(p[5])
```

## Genomic coordinates of the transcripts/exons

```{r etrid2grl, echo=TRUE, cache=TRUE}
grl <- etrid2grl(acols(p)$ENST)
pcgrl <- proteinCoding(grl)
```

```{r gviz1, echo=FALSE, cache=TRUE}
grTrack <- GeneRegionTrack(grl[[5]],
                           name = acols(p)$ENST[5])
grTrack2 <- GeneRegionTrack(pcgrl[[5]],
                           name = acols(p)$ENST[5])
ideoTrack <- IdeogramTrack(genome = "hg38",
                           chromosome = "chrX")
axisTrack <- GenomeAxisTrack()
plotTracks(list(ideoTrack, axisTrack, grTrack, grTrack2),
           add53 = TRUE, add35 = TRUE)

```
