# Mapping peptides to genomic coordinates

Illustration with [`Pbase`](http://bioconductor.org/packages/devel/bioc/html/Pbase.html).

```{r env, echo=FALSE}
suppressPackageStartupMessages(library("Pbase"))
library("knitr")
data(p)
```

We have an example data composed of `r length(p)` proteins, with
UniProt accession numbers and Ensembl transcipt identifers:

```{r ids}
kable(cbind(Acc = seqnames(p), ENST = acols(p)$ENST))
```

And each of these proteins have a set experimentally observed
peptides:

```{r, echo=FALSE}
kable(rbind(elementLengths(pranges(p))))
```

```{r pplot1, echo=FALSE}
plot(p[5])
```

## Genomic coordinates of the transcripts/exons

```{r etrid2grl, echo=TRUE}
grl <- etrid2grl(acols(p)$ENST)
pcgrl <- proteinCoding(grl)
```

```{r gviz1, echo=FALSE}
grTrack <- GeneRegionTrack(grl[[5]],
                           name = acols(p)$ENST[5])
grTrack2 <- GeneRegionTrack(pcgrl[[5]],
                           name = acols(p)$ENST[5])
ideoTrack <- IdeogramTrack(genome = "hg38",
                           chromosome = "chrX")
axisTrack <- GenomeAxisTrack()
plotTracks(list(ideoTrack, axisTrack, grTrack, grTrack2),
           add53 = TRUE, add35 = TRUE)

```
