# Mapping MS peptides and RNA-Seq short reads

```{r env, echo=FALSE}
library("knitr")
opts_chunk$set(fig.width=12, fig.height=7)
```

The last step of the mapping process is the combine the newly mapped
peptides and reads from RNA-Seq experiments. The figures below
illustrate this with data from Sheynkman et
al. ([2013](http://www.ncbi.nlm.nih.gov/pubmed/23629695),
[2014](http://www.ncbi.nlm.nih.gov/pubmed/25149441)) from the Jurkat
cell line (TIB-152). The
[mass spectrometry](https://db.systemsbiology.net/sbeams/cgi/PeptideAtlas/PASS_View?identifier=PASS00215)
and
[RNA-Seq](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE45428)
(SRR791580) where processed with standard pipelines. 


```{r peptrack,echo=FALSE}
suppressPackageStartupMessages(library("Gviz"))
load("./data/peptrack.rda")
plotTracks(peptrack[-5])
plotTracks(peptrack)
```

```{r alntrack}
options(ucscChromosomeNames=FALSE)
load("./data/alntrack.rda")
alntrack[[2]]@reference <- "./data/mybam.bam"
plotTracks(alntrack, from = 30116807, to = 30186474)
```

### References

Laurent Gatto and Sebastian Gibb
(2014). [`Pbase`](http://bioconductor.org/packages/devel/bioc/html/Pbase.html):
Manipulating and exploring protein and proteomics data. R package
version 0.6.9. https://github.com/ComputationalProteomicsUnit/Pbase

Pang et al. Tools to covisualize and coanalyze proteomic data with
genomes and transcriptomes: validation of genes and alternative mRNA
splicing. J Proteome Res. 2014 Jan 3;13(1):84-98. doi:
10.1021/pr400820p. Epub 2013 Nov 12. PubMed
[PMID: 24152167](http://www.ncbi.nlm.nih.gov/pubmed/24152167).

