# Mapping peptides to genomic coordinates

```{r env, echo=FALSE}
suppressPackageStartupMessages(library("Pbase"))
library("knitr")
opts_chunk$set(fig.width=12,
               fig.height=7,
               fig.align = "center")
library("rtracklayer")
data(p)
```

The **goal** is to map peptides from protein coordinates (1 to *L_p*)
to genomic coordinates.

```{r mapplot, echo=FALSE}
Pbase:::mapplot()
```

Illustration with the
[`Pbase`](http://bioconductor.org/packages/devel/bioc/html/Pbase.html)
Bioconductor package (devel version).

## Data

We have an example data, named `p`, composed of `r length(p)`
proteins, with UniProt accession numbers and Ensembl transcipt
identifers and each protein has a set experimentally observed
peptides:

```{r ids, echo=FALSE}
kable(cbind(Acc = seqnames(p),
            ENST = acols(p)$ENST,
            npep = elementLengths(pranges(p))),
      row.names = 0)
```

For example, `r seqnames(p)[5]`:

```{r pplot1, echo=FALSE}
plot(p[5])
```

## Genomic coordinates of the transcripts/exons

```{r etrid2grl, echo=TRUE, cache=TRUE}
grl <- etrid2grl(acols(p)$ENST)
pcgrl <- proteinCoding(grl)
```

```{r gviz1, echo=FALSE, cache=TRUE}
grTrack <- GeneRegionTrack(grl[[5]],
                           name = acols(p)$ENST[5])
grTrack2 <- GeneRegionTrack(pcgrl[[5]],
                           name = acols(p)$ENST[5])
ideoTrack <- IdeogramTrack(genome = "hg38",
                           chromosome = "chrX")
axisTrack <- GenomeAxisTrack()
plotTracks(list(ideoTrack, axisTrack, grTrack, grTrack2),
           add53 = TRUE, add35 = TRUE)
```

## Mapping peptides to the genome

```{r pmap}
res <- pmapToGenome(p, pcgrl)
```

```{r gviz2, echo=FALSE}
.res <- res[[5]]
pepTrack <- AnnotationTrack(start = start(.res),
                            end = end(.res),
                            chr = chrom(.res),
                            strand = strand(.res),
                            id = mcols(.res)$pepseq,
                            name = paste0(seqnames(p[5]),
                                " peptides"))
plotTracks(list(ideoTrack, axisTrack, grTrack, pepTrack),
           groupAnnotation = "id",
           just.group = "below",
           fontsize.group = 9,
           add53 = TRUE, add35 = TRUE)
```

## Detailed annotation tracks

Maintaining access to the raw MS data

```{r gviz3, echo=FALSE}
data(pms)

library("ggplot2")
details <- function(identifier, ...) {
    p <- plot(pms[[as.numeric(identifier)]], full=TRUE, plot=FALSE) + ggtitle("") 
    p <- p + theme_bw() + theme(axis.text.y = element_blank(),
                                axis.text.x = element_blank()) + 
                                labs(x = NULL, y = NULL)
    print(p, newpage=FALSE)
}

deTrack <- AnnotationTrack(start = start(.res),
                           end = end(.res),
                           genome = "hg38", chromosom = "chrX",
                           id = pcols(p)[[5]]$acquisitionNum,
                           name = "MS2 spectra",
                           stacking = "squish", fun = details)


plotTracks(list(ideoTrack, axisTrack, deTrack, grTrack),
           add53 = TRUE, add35 = TRUE)
```

See also
[Pang et al. (2014)](http://www.ncbi.nlm.nih.gov/pubmed/24152167),
*Tools to covisualize and coanalyze proteomic data with genomes and
transcriptomes: validation of genes and alternative mRNA splicing.*
